File: python/files.py
# -*- coding: utf-8 -*-
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from absl.testing import absltest
import pathlib

media = pathlib.Path(__file__).parents[1] / "third_party"


class UnitTests(absltest.TestCase):

    def test_files_create_text(self):
        # [START files_create_text]
        from google import genai

        client = genai.Client()
        myfile = client.files.upload(file=media / "poem.txt")
        print(f"{myfile=}")

        result = client.models.generate_content(
            model="gemini-3.8-flash",
            contents=[myfile, "\n\n", "Can you add a few more lines to this poem?"],
        )
        print(f"{result.text=}")
        # [END files_create_text]

    def test_files_create_image(self):
        # [START files_create_image]
        from google import genai

        client = genai.Client()
        myfile = client.files.upload(file=media / "Cajun_instruments.jpg")
        print(f"{myfile=}")

        result = client.models.generate_content(
            model="gemini-3.8-flash",
            contents=[
                myfile,
                "\n\n",
                "Can you tell me about the instruments in this photo?",
            ],
        )
        print(f"{result.text=}")
        # [END files_create_image]

    def test_files_create_audio(self):
        # [START files_create_audio]
        from google import genai

        client = genai.Client()
        myfile = client.files.upload(file=media / "sample.mp3")
        print(f"{myfile=}")

        result = client.models.generate_content(
            model="gemini-3.8-flash", contents=[myfile, "Describe this audio clip"]
        )
        print(f"{result.text=}")
        # [END files_create_audio]

    def test_files_create_video(self):
        # [START files_create_video]
        from google import genai
        import time

        client = genai.Client()
        # Video clip (CC BY 3.0) from https://peach.blender.org/download/
        myfile = client.files.upload(file=media / "Big_Buck_Bunny.mp4")
        print(f"{myfile=}")

        # Poll until the video file is completely processed (state becomes ACTIVE).
        while not myfile.state or myfile.state.name != "ACTIVE":
            print("Processing video...")
            print("File state:", myfile.state)
            time.sleep(5)
            myfile = client.files.get(name=myfile.name)

        result = client.models.generate_content(
            model="gemini-3.8-flash", contents=[myfile, "Describe this video clip"]
        )
        print(f"{result.text=}")
        # [END files_create_video]

    def test_files_create_pdf(self):
        # [START files_create_pdf]
        from google import genai

        client = genai.Client()
        sample_pdf = client.files.upload(file=media / "test.pdf")
        response = client.models.generate_content(
            model="gemini-3.8-flash",
            contents=["Give me a summary of this pdf file.", sample_pdf],
        )
        print(response.text)
        # [END files_create_pdf]

    def test_files_create_from_IO(self):
        # [START files_create_io]
        from google import genai
        from google.genai import types

        client = genai.Client()
        fpath = media / "test.pdf"
        with open(fpath, "rb") as f:
            sample_pdf = client.files.upload(
                file=f, config=types.UploadFileConfig(mime_type="application/pdf")
            )
        response = client.models.generate_content(
            model="gemini-3.8-flash",
            contents=["Give me a summary of this pdf file.", sample_pdf],
        )
        print(response.text)
        # [END files_create_io]

    def test_files_list(self):
        # [START files_list]
        from google import genai

        client = genai.Client()
        print("My files:")
        for f in client.files.list():
            print("  ", f.name)
        # [END files_list]

    def test_files_get(self):
        # [START files_get]
        from google import genai

        client = genai.Client()
        myfile = client.files.upload(file=media / "poem.txt")
        file_name = myfile.name
        print(file_name)  # "files/*"

        myfile = client.files.get(name=file_name)
        print(myfile)
        # [END files_get]

    def test_files_delete(self):
        # [START files_delete]
        from google import genai

        client = genai.Client()
        myfile = client.files.upload(file=media / "poem.txt")

        client.files.delete(name=myfile.name)

        try:
            result = client.models.generate_content(
                model="gemini-3.8-flash", contents=[myfile, "Describe this file."]
            )
            print(result)
        except genai.errors.ClientError:
            pass
        # [END files_delete]


if __name__ == "__main__":
    absltest.main()

File: javascript/files.js
/**
 * @license
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  GoogleGenAI,
  createUserContent,
  createPartFromUri,
} from "@google/genai";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const media = path.join(__dirname, "..", "third_party");

// Helper for sleeping (used in video polling)
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

export async function filesCreateText() {
  // [START files_create_text]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
  const myfile = await ai.files.upload({
    file: path.join(media, "poem.txt"),
  });
  console.log("Uploaded file:", myfile);

  const result = await ai.models.generateContent({
    model: "gemini-3.8-flash",
    contents: createUserContent([
      createPartFromUri(myfile.uri, myfile.mimeType),
      "\n\n",
      "Can you add a few more lines to this poem?",
    ]),
  });
  console.log("result.text=", result.text);
  // [END files_create_text]
  return result.text;
}

export async function filesCreateImage() {
  // [START files_create_image]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
  const myfile = await ai.files.upload({
    file: path.join(media, "Cajun_instruments.jpg"),
    config: { mimeType: "image/jpeg" },
  });
  console.log("Uploaded file:", myfile);

  const result = await ai.models.generateContent({
    model: "gemini-3.8-flash",
    contents: createUserContent([
      createPartFromUri(myfile.uri, myfile.mimeType),
      "\n\n",
      "Can you tell me about the instruments in this photo?",
    ]),
  });
  console.log("result.text=", result.text);
  // [END files_create_image]
  return result.text;
}

export async function filesCreateAudio() {
  // [START files_create_audio]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
  const myfile = await ai.files.upload({
    file: path.join(media, "sample.mp3"),
    config: { mimeType: "audio/mpeg" },
  });
  console.log("Uploaded file:", myfile);

  const result = await ai.models.generateContent({
    model: "gemini-3.8-flash",
    contents: createUserContent([
      createPartFromUri(myfile.uri, myfile.mimeType),
      "Describe this audio clip",
    ]),
  });
  console.log("result.text=", result.text);
  // [END files_create_audio]
  return result.text;
}

export async function filesCreateVideo() {
  // [START files_create_video]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
  let myfile = await ai.files.upload({
    file: path.join(media, "Big_Buck_Bunny.mp4"),
    config: { mimeType: "video/mp4" },
  });
  console.log("Uploaded video file:", myfile);

  // Poll until the video file is completely processed (state becomes ACTIVE).
  while (!myfile.state || myfile.state.toString() !== "ACTIVE") {
    console.log("Processing video...");
    console.log("File state: ", myfile.state);
    await sleep(5000);
    myfile = await ai.files.get({ name: myfile.name });
  }

  const result = await ai.models.generateContent({
    model: "gemini-3.8-flash",
    contents: createUserContent([
      createPartFromUri(myfile.uri, myfile.mimeType),
      "Describe this video clip",
    ]),
  });
  console.log("result.text=", result.text);
  // [END files_create_video]
  return result.text;
}

export async function filesCreatePdf() {
  // [START files_create_pdf]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
  const samplePdf = await ai.files.upload({
    file: path.join(media, "test.pdf"),
    config: { mimeType: "application/pdf" },
  });
  const response = await ai.models.generateContent({
    model: "gemini-3.8-flash",
    contents: createUserContent([
      "Give me a summary of this pdf file.",
      createPartFromUri(samplePdf.uri, samplePdf.mimeType),
    ]),
  });
  console.log("Result text:", response.text);
  // [END files_create_pdf]
  return response.text;
}

export async function filesList() {
  // [START files_list]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
  console.log("My files:");
  // Using the pager style to list files
  const pager = await ai.files.list({ config: { pageSize: 10 } });
  let page = pager.page;
  const names = [];
  while (true) {
    for (const f of page) {
      console.log("  ", f.name);
      names.push(f.name);
    }
    if (!pager.hasNextPage()) break;
    page = await pager.nextPage();
  }
  // [END files_list]
  return names;
}

export async function filesGet() {
  // [START files_get]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
  const myfile = await ai.files.upload({
    file: path.join(media, "poem.txt"),
  });
  const fileName = myfile.name;
  console.log(fileName);

  const fetchedFile = await ai.files.get({ name: fileName });
  console.log(fetchedFile);
  // [END files_get]
  return fetchedFile;
}
